prov/sm2/src/sm2_coordination.c

```diff
@@ -0,0 +1,675 @@
+	return -FI_EAVAIL;
+
+found:
+
+	if (self) {
+		entries[jentry].startup_ready = 0;
+		atomic_wmb();
+		entries[jentry].pid = pid;
+	}
+
+	if (!self && entries[jentry].pid == 0) {
+		entries[jentry].startup_ready = 0;
+		atomic_wmb();
+		entries[jentry].pid = -pid;
+	}
+
+	FI_WARN(&sm2_prov, FI_LOG_AV, "Allocated sm2 region for %s at AV[%d]\n", name,
+		jentry);
+	strncpy(entries[jentry].ep_name, name, SM2_NAME_MAX);
+
+	*av_key = jentry;
+
+	/* With the entry allocated, we now need to ensure it's mapped. */
+	if (!sm2_mapping_long_enough_check(map, jentry)) {
+		sm2_coordinator_extend_for_entry(map, jentry);
+	}
+	return 0;
+}
+
+/**
+ * @brief look-up the name.
+ *
+ * @param[in] name
+ * @param[in] map
+ * @return the index of the name, or -1
+ */
+int sm2_coordinator_lookup_entry(const char *name, struct sm2_mmap *map)
+{
+
+	struct sm2_coord_file_header *header = (void *)map->base;
+	struct sm2_ep_allocation_entry *entries;
+	int jentry;
+
+	entries = sm2_mmap_entries(map);
+
+	for (jentry = 0; jentry < header->ep_enumerations_max; jentry++) {
+		if (0 == strncmp(name, entries[jentry].ep_name, SM2_NAME_MAX)) {
+			FI_WARN(&sm2_prov, FI_LOG_AV, "Found existing %s in slot %d\n",
+				name, jentry);
+			return jentry;
+		}
+		if (entries[jentry].ep_name[0] != '\0') {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Searching for %s. Not yet found in spot AV[%d]=%s, PID= "
+				"%d\n",
+				name, jentry, entries[jentry].ep_name,
+				entries[jentry].pid);
+		}
+		else if (!strcmp(entries[jentry].ep_name, ZOMBIE_ALLOCATION_NAME)) {
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Found zombie in spot AV[%d], PID= %d\n", jentry,
+				entries[jentry].pid);
+		}
+	}
+	return -1;
+}
+
+/**
+ * @brief clear the pid for this entry.  must already hold lock.
+ *
+ * @param[in] name
+ * @param[inout] map
+ * @param[out] av_key
+ */
+ssize_t sm2_coordinator_free_entry(struct sm2_mmap *map, int av_key)
+{
+	struct sm2_ep_allocation_entry *entries;
+	int pid = getpid();
+	(void)pid;
+	entries = sm2_mmap_entries(map);
+	assert(entries[av_key].pid == pid);
+	entries[av_key].pid = 0;
+	return 0;
+}
+
+ssize_t sm2_coordinator_lock(struct sm2_mmap *map)
+{
+	struct sm2_coord_file_header *header = (void *)map->base;
+	int pid = getpid();
+
+	pthread_mutex_lock(&header->write_lock);
+	ofi_atomic_set32(&header->pid_lock_hint, pid);
+	return 0;
+}
+
+ssize_t sm2_coordinator_unlock(struct sm2_mmap *map)
+{
+	struct sm2_coord_file_header *header = (void *)map->base;
+	int pid = getpid();
+	int hint_pid = ofi_atomic_get32(&header->pid_lock_hint);
+	if (pid != hint_pid) {
+		FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+			"We should (%d) hold the lock, but pid_lock_hint is %d!", pid,
+			hint_pid);
+	}
+	pthread_mutex_unlock(&header->write_lock);
+
+	/* use CAS to clear pid because it's possible another process has
+	   already acquired the lock.  CAS will only clear it if we hold it. */
+	ofi_atomic_cas_bool32(&header->pid_lock_hint, pid, 0);
+	return 0;
+}
```

Comment:
I would be interested in knowing the overhead of having all of this hint code inside the locking/unlocking path. I see that it is needed for clearing up the mapping if the peer has died while holding the lock but I think this has the potential to introduce more overhead than it's worth. We have signal handlers to clean up a peer if it does die. I'm wondering if we can use something there to handle the case that this protects against?

Reply:
this is also startup only, and the lock is only used for peers that wish to modify the SHM file header (i.e.. insert themselves into header)... not saying it shouldn't be fixed, but it is lower priority.  I think we care more about startup correctness than performance.